ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java

```diff
@@ -799,17 +963,33 @@ private void processEntryForLogSegmentBucket(ScanResults results, IndexEntry ind
     IndexValue indexValue = indexEntry.getValue();
     results.updateLogSegmentBaseBucket(indexValue.getOffset().getName(), indexValue.getSize());
     if (indexValue.isPut()) {
-      long expOrDelTimeInMs = indexValue.getExpiresAtMs();
+      long expiresAtMs = indexValue.getExpiresAtMs();
+      long deleteAtMs = Utils.Infinite_Time;
       IndexFinalState finalState = keyFinalStates.get(indexEntry.getKey());
       if (finalState != null && finalState.isDelete()) {
-        long deleteTimeInMs = finalState.getOperationTime();
-        expOrDelTimeInMs =
-            expOrDelTimeInMs != Utils.Infinite_Time && expOrDelTimeInMs < deleteTimeInMs ? expOrDelTimeInMs
-                : deleteTimeInMs;
+        deleteAtMs = finalState.getOperationTime();
+      }
+      if (expiresAtMs == Utils.Infinite_Time && deleteAtMs != Utils.Infinite_Time) {
+        handleLogSegmentDeletedBucketUpdate(results, indexValue, deleteAtMs, SUBTRACT);
+      } else if (expiresAtMs != Utils.Infinite_Time && deleteAtMs == Utils.Infinite_Time) {
+        handleLogSegmentExpiredBucketUpdate(results, indexValue, expiresAtMs, SUBTRACT);
+      } else if (expiresAtMs != Utils.Infinite_Time && deleteAtMs != Utils.Infinite_Time) {
+        if (expiresAtMs < deleteAtMs) {
+          handleLogSegmentExpiredBucketUpdate(results, indexValue, expiresAtMs, SUBTRACT);
+        } else {
+          handleLogSegmentDeletedBucketUpdate(results, indexValue, deleteAtMs, SUBTRACT);
+        }
       }
-      if (!indexValue.isTtlUpdate() || !isTtlUpdateEntryValid(indexEntry.getKey(), indexValue,
-          expOrDelTimeInMs == Utils.Infinite_Time ? expOrDelTimeInMs : expOrDelTimeInMs + 1, keyFinalStates)) {
-        handleLogSegmentBucketUpdate(results, indexValue, expOrDelTimeInMs, SUBTRACT);
+    } else if (!indexValue.isDelete() && !indexValue.isUndelete() && indexValue.isTtlUpdate()) {
+      IndexFinalState finalState = keyFinalStates.get(indexEntry.getKey());
+      if (finalState != null && finalState.isDelete()) {
+        IndexValue putValue =
+            index.findKey(indexEntry.getKey(), new FileSpan(index.getStartOffset(), indexValue.getOffset()),
+                EnumSet.of(PersistentIndex.IndexEntryType.PUT));
```

Comment:
I understand the piece of code is to check whether TtlUpdate is valid or not. If PUT and TtlUpdate are in same log segment and final state is delete, then TtlUpdate is not valid. 
I wonder if we can search the log segment of TtlUpdate only instead of findKey from the very beginning of whole PersistentIndex.

Reply:
that's a good idea, updated.